plugins/outputs/pulsar/pulsar.go

```diff
@@ -0,0 +1,378 @@
+		if err != nil {
+			p.Log.Error("Invalid ConnectionTimeout. Setting it to 30s ", err.Error())
+			clientOptions.ConnectionTimeout = 30 * time.Second
+		} else {
+			clientOptions.ConnectionTimeout = connectionTimeout
+		}
+	}
+	if p.OperationTimeout != "" {
+		operationTimeout, err := time.ParseDuration(p.OperationTimeout)
+		if err != nil {
+			p.Log.Error("Invalid OperationTimeout. Setting it to 30s", err.Error())
+			clientOptions.OperationTimeout = 30 * time.Second
+		} else {
+			clientOptions.OperationTimeout = operationTimeout
+		}
+
+	}
+
+	if p.AuthProvider == "tls" {
+		if p.TLSCertificatePath != "" && p.TLSPrivateKeyPath != "" {
+			clientOptions.Authentication = pulsar.NewAuthenticationTLS(p.TLSCertificatePath, p.TLSPrivateKeyPath)
+		} else {
+			return fmt.Errorf("tls auth specified but tls_certificate_path or tls_private_key_path is miising")
+		}
+
+	} else if p.AuthProvider == "token" {
+		if p.AuthToken != "" {
+			clientOptions.Authentication = pulsar.NewAuthenticationToken(p.AuthToken)
+		} else {
+			return fmt.Errorf("auth provider is token but auth_token is empty")
+		}
+
+	}
+
+	client, err := pulsar.NewClient(clientOptions)
+	if err != nil {
+		return err
+	}
+	p.client = client
+	return nil
+}
+
+func (p *Pulsar) Close() error {
+	if p.client != nil {
+		p.client.Close()
+	}
+	return nil
+}
+
+func (p *Pulsar) SampleConfig() string {
+	return sampleConfig
+}
+
+func (p *Pulsar) Description() string {
+	return "Configuration to send metrics to pulsar broker"
+}
+
+func (p *Pulsar) routingKey(metric telegraf.Metric) string {
+	if p.RoutingTag != "" {
+		key, ok := metric.GetTag(p.RoutingTag)
+		if ok {
+			return key
+		}
+	}
+
+	if p.RoutingKey == "measurement" {
+		return metric.Name()
+	}
+
+	if p.RoutingKey == "random" {
+		u, err := uuid.NewV4()
+		if err != nil {
+			p.Log.Errorf("Unable to generate uudid %s", err.Error())
+			return ""
+		}
+		return u.String()
+	}
+	return p.RoutingKey
+}
+func sendAsyncCallback(id pulsar.MessageID, message *pulsar.ProducerMessage, err error) {
+	if err != nil {
+		log.Printf("Failed to publish metrics : %s", err.Error())
+	}
+
+}
+
+func (p *Pulsar) Write(metrics []telegraf.Metric) error {
+	var batchSize uint
+	// Setting the batch size to metric_batch_size is unset
+	if p.BatchingMaxMessages == 0 {
+		batchSize = uint(len(metrics))
+	} else {
+		batchSize = p.BatchingMaxMessages
+	}
+	for _, metric := range metrics {
+		metric, topic := p.GetTopicName(metric)
+		producer, err := p.GetProducer(topic, batchSize)
+		if err == nil {
+			buf, err := p.serializer.Serialize(metric)
+			if err != nil {
+				p.Log.Debugf("Could not serialize metric: %v", err)
+				continue
+			}
+			m := pulsar.ProducerMessage{
+				Payload:   buf,
+				EventTime: metric.Time(),
+				Key:       p.routingKey(metric),
+			}
+
+			// Negative timestamps are not allowed by the Pulsar protocol.
+			if !metric.Time().Before(zeroTime) {
+				m.EventTime = metric.Time()
+			}
```

Comment:
Well even though it is not allowed you still use that time? Did you mean
```suggestion
			if !metric.Time().Before(zeroTime) {
				m.EventTime = time.Now()
			}
```